resolwe/flow/models.py

```diff
@@ -583,6 +594,81 @@ def validation_schema(name):
 
 
+TYPE_SCHEMA = validation_schema('type')
+
+
+def validate_schema(instance, schema):
+    """Check if DictField values are consistent with our data types.
+
+    Perform basic JSON schema validation and our custom validations:
+
+      * check that required fields are given
+      * check if ``basic:file:`` and ``list:basic:file`` fields matches
+        regex given in schema (only if ``validate_regex`` is defined in
+        schema for coresponding fields)
+      * check that referenced ``Data`` objects (in ``data:<data_type>``
+        and  ``list:data:<data_type>`` fields) exists and are of type
+        ``<data_type>``
+      * check that referenced ``Storage`` objects (in ``basic:json``
+        fields) exists
+
+    :param list instance: Instance to be validated
+    :param list schema: Schema for validation
+    :rtype: None
+    :raises ValidationError: if ``instance`` doesn't match schema
+      defined in ``schema``
+
+    """
+    def validate_file(filename, reg):
+        """Validate file name."""
+        if not re.search(reg, filename):
+            raise ValidationError(
+                "File name {} does not match regex {}".format(filename, reg))
+
+    def validate_data(data_pk, type_):
+        """"Check that `Data` objects exist and is of right type."""
+        data_qs = Data.objects.filter(pk=field).values('process__type')
+        if not data_qs.exists():
+            raise ValidationError(
+                "Referenced `Data` object does not exist (id:{})".format(data_pk))
+        if not data_qs.first()['process__type'].startswith(type_):
+            raise ValidationError(
+                "Referenced `Data` object is of wrong type (id:{})".format(data_pk))
+
+    for _schema, _fields, _ in iterate_schema(instance, schema):
+        name = _schema['name']
+
+        if _schema.get('required', False) and name not in _fields:
+            raise ValidationError("Required field \"{}\" not given.".format(name))
+
+        if name in _fields:
+            field = _fields[name]
+            type_ = _schema.get('type', "")
+
+            try:
+                jsonschema.validate([{"type": type_, "value": field}], TYPE_SCHEMA)
+            except jsonschema.exceptions.ValidationError as ex:
+                raise ValidationError(ex.message)
+
+            if type_ == 'basic:file:' and 'validate_regex' in _schema:
+                validate_file(field['file'], _schema['validate_regex'])
+
+            if type_ == 'list:basic:file:' and 'validate_regex' in _schema:
+                for file_ in field:
+                    validate_file(file_['file'], _schema['validate_regex'])
+
+            if type_ == 'basic:json:' and not Storage.objects.filter(pk=field).exists():
+                raise ValidationError(
+                    "Referenced `Storage` object does not exist (id:{})".format(field))
+
+            if type_.startswith('data:'):
+                validate_data(field, type_[5:])  # remove "data:" from `type_`
```

Comment:
We should probably allow Data objects that do not exist, but they should be of the right type if they are present. Say you run a differential expression, 10 cases vs 10 controls. Later you remove some of the input data. The differential expression result is still valid. You can not access the original objects any more, but you probably do not have to. If you would remove the Data object from the list (to make it valid), you could get a false impression that the differential expression was only done on the subset of data. This should be handled correctly in UI: we should write "(deleted)" next to the link to the data objects that do not exist. And they should not be links, only labels.